File: mvvm-viewbinding/src/main/java/com/everdeng/android/lib/mvvm/viewbinding/base/BaseMvvmActivity.java
package com.everdeng.android.lib.mvvm.viewbinding.base;

import android.os.Bundle;
import android.view.LayoutInflater;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.viewbinding.ViewBinding;


import com.everdeng.android.lib.core.base.AbstractActivity;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;


/**
 * @author allever
 */
public abstract class BaseMvvmActivity<VB extends ViewBinding, VM extends ViewModel> extends AbstractActivity {

    protected VB mBinding;
    protected VM mViewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = bindView(getLayoutInflater());
        setContentView(mBinding.getRoot());
        Class modelClass;
        Type type = getClass().getGenericSuperclass();
        if (type instanceof ParameterizedType) {
            modelClass = (Class) ((ParameterizedType) type).getActualTypeArguments()[1];
        } else {
            modelClass = BaseViewModel.class;
        }

        mViewModel = (VM) new ViewModelProvider(this, new ViewModelProvider.Factory() {
            @NonNull
            @Override
            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
                try {
                    return modelClass.getDeclaredConstructor().newInstance();
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                } catch (InstantiationException e) {
                    e.printStackTrace();
                } catch (InvocationTargetException e) {
                    e.printStackTrace();
                } catch (NoSuchMethodException e) {
                    e.printStackTrace();
                }

                return null;
            }
        }).get(modelClass);

        observeData();
    }


    /**
     * @return
     */
    protected abstract VB bindView(LayoutInflater inflater);


    /**
     *
     */
    protected abstract void observeData();


}

File: mvvm/src/main/java/com/everdeng/android/lib/mvvm/base/BaseMvvmActivity.kt
package com.everdeng.android.lib.mvvm.base

import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.everdeng.android.lib.core.app.App.Companion.context
import com.everdeng.android.lib.core.base.AbstractActivity
import java.lang.reflect.ParameterizedType

abstract class BaseMvvmActivity<VM : ViewModel, DB : ViewDataBinding> : AbstractActivity() {

    protected lateinit var mBinding: DB
    protected lateinit var mViewModel: VM

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mBinding = DataBindingUtil.setContentView(this, getLayoutId())
        mBinding.lifecycleOwner = this
        mViewModel =
            ViewModelProvider(this, object : ViewModelProvider.Factory {
                override fun <VM : ViewModel?> create(modelClass: Class<VM>): VM {
                    return modelClass.getDeclaredConstructor().newInstance()
//                    val clz = (this@BaseMvvmActivity.javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<VM>
//                    return clz.getDeclaredConstructor().newInstance()
                }
            }).get((this.javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<VM>)
        mBinding.setVariable(getBindingVariable(), mViewModel)
        (mViewModel as BaseViewModel).bind(context, this)
        initView()
        (mViewModel as BaseViewModel).init()
    }

    override fun onDestroy() {
        super.onDestroy()
        (mViewModel as BaseViewModel).destroy()
    }

    abstract fun getLayoutId(): Int
    abstract fun getBindingVariable(): Int
    abstract fun initView()
}